### PR TITLE
Support quantized attributes in AccessorUtility

### PR DESCRIPTION
**Pull request by lilleyse**
## Description

Updates `AccessorUtility` to support quantized positions, normals, and texcoords enabled by [KHR_mesh_quantization](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_mesh_quantization/README.md).

## Author checklist

- [x] I have submitted a [Contributor License Agreement](https://github.com/CesiumGS/community/tree/main/CLAs) (only needed once).
- [x] I have done a full self-review of my code.
- [x] I have updated `CHANGES.md` with a short summary of my change (for user-facing changes).
- [x] I have added or updated unit tests to ensure consistent code coverage as necessary.
- [x] I have updated the documentation as necessary.

## Reviewer checklist

Thank you for taking the time to review this PR. By approving a PR you are taking as much responsibility for these changes as the author.

As you review, please go through the checklist below:

- [x] Review and run all parts of the test plan on this branch and verify it matches expectations.
    - [ ] If the issue is a bug please make sure you can reproduce the bug in the main branch and then checkout this branch to make sure it actually solved the issue.
- [x] Review the code and make sure you do not have any remaining questions or concerns. You should understand the code change and the chosen approach. If you are not confident or have doubts about the code, please do not hesitate to ask questions.
    - [x] Code should follow the [C++ Style Guide](https://cesium.com/learn/cesium-native/ref-doc/style-guide.html)
- [x] Review the unit tests and make sure there are no missing tests or edge cases.
- [x] Review documentation changes and updates to `CHANGES.md` to make sure they accurately cover the work in this PR.
- [x] Verify that the [Contributor License Agreement](https://github.com/CesiumGS/community/tree/main/CLAs) has been submitted, if needed.

### CODE REVIEW

File: Cesium3DTilesContent/src/I3dmToGltfConverter.cpp [hunk 1]
```diff
@@ -734,7 +734,7 @@ std::vector<glm::dmat4> getMeshGpuInstancingTransforms(
     }
   }
   if (rotations) {
-    auto quatAccessorView = getQuaternionAccessorView(model, rotations);
+    auto quatAccessorView = getQuaternionAccessorView(model, *rotations);
```
Comment (j9liu) [inline Cesium3DTilesContent/src/I3dmToGltfConverter.cpp]: Not your code, but can we make the `translations`/`rotations`/`scales` pointers be properly prefixed by `p`?

File: CHANGES.md [hunk 2]
```diff
@@ -5,10 +5,18 @@
 ##### Breaking Changes :mega:
 
 - Upgraded vcpkg to `2026.04.27`.
+- `PositionAccessorType` and `NormalAccessorType` are now variant types to support quantized attributes enabled by `KHR_mesh_quantization`.
+- The `AccessorView` constructor overload that takes raw buffer data now requires an additional `normalized` argument.
+- The `getQuaternionAccessorView` overload that takes an accessor pointer now takes an accessor reference.
```
Comment (j9liu) [inline CHANGES.md]: **NIT**

```suggestion
- The `getQuaternionAccessorView` overload that took an accessor pointer now takes an accessor reference.
```

File: CesiumGltf/include/CesiumGltf/AccessorUtility.h [hunk 3]
```diff
@@ -58,9 +79,51 @@ PositionAccessorType
...
+ *
+ * There are technically no invalid position values, so we use std::nullopt to
+ * denote an erroneous value.
+ */
+struct PositionFromAccessor {
+  /**
+   * @brief Attempts to obtain a `glm::dvec3` at the given index from an
+   * accessor over a vec3. The values will be cast to `double` and normalized
+   * based on `std::numeric_limits<T>::max()`. If the index is invalid,
+   * `std::nullopt` is returned instead.
```
Comment (j9liu) [inline CesiumGltf/include/CesiumGltf/AccessorUtility.h]: **NIT**:
```suggestion
   * @brief Attempts to obtain a `glm::dvec3` at the given index from an
   * accessor over a vec3. The values will be cast to `double` and, if applicable,
   * normalized based on `std::numeric_limits<T>::max()`. If the index is invalid,
   * `std::nullopt` is returned instead.
```

File: CesiumGltf/include/CesiumGltf/AccessorUtility.h [hunk 4]
```diff
@@ -58,9 +79,51 @@ PositionAccessorType
 getPositionAccessorView(const Model& model, const MeshPrimitive& primitive);
 
 /**
- * Type definition for normal accessor.
+ * Visitor that retrieves the position from the given accessor type
+ * as a glm::dvec3.
+ *
+ * There are technically no invalid position values, so we use std::nullopt to
+ * denote an erroneous value.
```
Comment (j9liu) [inline CesiumGltf/include/CesiumGltf/AccessorUtility.h]: **NIT**
```suggestion
 * Visitor that retrieves the position from the given accessor type
 * as a `glm::dvec3`.
 *
 * There are technically no invalid position values, so `std::nullopt` is used to
 * denote an erroneous value.
```

File: CesiumGltf/include/CesiumGltf/AccessorUtility.h [hunk 5]
```diff
@@ -70,6 +133,44 @@ typedef AccessorView<AccessorTypes::VEC3<float>> NormalAccessorType;
 NormalAccessorType
 getNormalAccessorView(const Model& model, const MeshPrimitive& primitive);
 
+/**
+ * Visitor that retrieves the normal from the given accessor type
+ * as a glm::dvec3.
+ *
+ * There are technically no invalid normal values, so we use std::nullopt to
+ * denote an erroneous value.
```
Comment (j9liu) [inline CesiumGltf/include/CesiumGltf/AccessorUtility.h:141]: **NIT**
```suggestion
 * Visitor that retrieves the normal from the given accessor type
 * as a `glm::dvec3`.
 *
 * There are technically no invalid normal values, so `std::nullopt` is used to
 * denote an erroneous value.
```

File: CesiumGltf/include/CesiumGltf/AccessorUtility.h [hunk 6]
```diff
@@ -70,6 +133,44 @@ typedef AccessorView<AccessorTypes::VEC3<float>> NormalAccessorType;
...
+ */
+struct NormalFromAccessor {
+  /**
+   * @brief Attempts to obtain a `glm::dvec3` at the given index from an
+   * accessor over a vec3. The values will be cast to `double` and normalized
+   * based on `std::numeric_limits<T>::max()`. If the index is invalid,
+   * `std::nullopt` is returned instead.
```
Comment (j9liu) [inline CesiumGltf/include/CesiumGltf/AccessorUtility.h]: **NIT**:
```suggestion
   * @brief Attempts to obtain a `glm::dvec3` at the given index from an
   * accessor over a vec3. The values will be cast to `double` and, if applicable,
   * normalized based on `std::numeric_limits<T>::max()`. If the index is invalid,
   * `std::nullopt` is returned instead.
```

File: CesiumGltf/include/CesiumGltf/AccessorUtility.h [hunk 7]
```diff
@@ -365,20 +468,6 @@ TexCoordAccessorType getTexCoordAccessorView(
  * behavior, so we use std::nullopt to denote an erroneous value.
```
Comment (j9liu) [inline CesiumGltf/include/CesiumGltf/AccessorUtility.h]: Ah, so this is the original comment that the others were copied from. Still worth changing for consistency...

```suggestion
* Visitor that retrieves the texture coordinates from the given accessor type
* as a `glm::dvec2`. This should be initialized with the target index.
*
* There are technically no invalid UV values because of clamp / wrap
* behavior, so `std::nullopt` is used to denote an erroneous value.
```

File: CesiumGltf/include/CesiumGltf/AccessorUtility.h [hunk 8]
```diff
@@ -411,37 +499,76 @@ struct TexCoordFromAccessor {
...
  * the given \ref Model.
  *
- * @param model The model containing the quaternion.
+ * @param model The model containing the quaternion accessor.
  * @param accessor An accessor from which the quaternion will be obtained.
- * @returns A quaternion from the data in `accessor`. If no quaternion could be
- * obtained, the default value for \ref QuaternionAccessorType will be returned
- * instead.
+ * @returns A quaternion accessor view from the data in `accessor`. If no
+ * quaternion could be obtained, the default value for \ref
```
Comment (j9liu) [inline CesiumGltf/include/CesiumGltf/AccessorUtility.h]:
```suggestion
 * @returns A quaternion accessor view from the data in `accessor`. If no
 * such view could be obtained, the default value for \ref
```

File: CesiumGltf/include/CesiumGltf/AccessorUtility.h [hunk 9]
```diff
@@ -411,37 +499,76 @@ struct TexCoordFromAccessor {
...
- * @param model The model containing the quaternion.
+ * @param model The model containing the quaternion accessor.
  * @param accessorIndex An index to the accessor from which the quaternion will
  * be obtained.
- * @returns A quaternion from the data in the accessor at `accessorIndex`. If no
- * quaternion could be obtained, the default value for \ref
- * QuaternionAccessorType will be returned instead.
+ * @returns A quaternion accessor view from the data in the accessor at
+ * `accessorIndex`. If no quaternion could be obtained, the default value for
+ * \ref QuaternionAccessorType will be returned instead.
```
Comment (j9liu) [inline CesiumGltf/include/CesiumGltf/AccessorUtility.h:531]:
```suggestion
 * @returns A quaternion accessor view from the data in the accessor at
 * `accessorIndex`. If no such view could be obtained, the default value for
 * \ref QuaternionAccessorType will be returned instead.
```

File: CesiumGltf/include/CesiumGltf/AccessorUtility.h [hunk 10]
```diff
@@ -411,37 +499,76 @@ struct TexCoordFromAccessor {
...
+/**
+ * Visitor that retrieves the quaternion from the given accessor type
+ * as a glm::dquat.
+ */
+struct QuaternionFromAccessor {
+  /**
+   * @brief Attempts to obtain a `glm::dquat` at the given index from an
+   * accessor over a vec4. The values will be cast to `double` and normalized
+   * based on `std::numeric_limits<T>::max()`. If the index is invalid,
+   * `std::nullopt` is returned instead.
```
Comment (j9liu) [inline CesiumGltf/include/CesiumGltf/AccessorUtility.h]: **NIT**
```suggestion
   * @brief Attempts to obtain a `glm::dquat` at the given index from an
   * accessor over a vec4. The values will be cast to `double` and, if applicable,
   * normalized based on `std::numeric_limits<T>::max()`. If the index is invalid,
   * `std::nullopt` is returned instead.
```

File: CesiumGltf/src/AccessorUtility.cpp [hunk 11]
```diff
@@ -19,12 +19,24 @@ getPositionAccessorView(const Model& model, const MeshPrimitive& primitive) {
...
+  case Accessor::ComponentType::BYTE:
+    return AccessorView<AccessorTypes::VEC3<int8_t>>(model, *pAccessor);
+  case Accessor::ComponentType::UNSIGNED_BYTE:
+    return AccessorView<AccessorTypes::VEC3<uint8_t>>(model, *pAccessor);
+  case Accessor::ComponentType::SHORT:
+    return AccessorView<AccessorTypes::VEC3<int16_t>>(model, *pAccessor);
+  case Accessor::ComponentType::UNSIGNED_SHORT:
+    return AccessorView<AccessorTypes::VEC3<uint16_t>>(model, *pAccessor);
+  case Accessor::ComponentType::FLOAT:
+    return AccessorView<AccessorTypes::VEC3<float>>(model, *pAccessor);
```
Comment (j9liu) [inline CesiumGltf/src/AccessorUtility.cpp:36]: This doesn't check for `normalized` but some of the other functions do -- is this for any particular reason? Should they be consistent?

File: CHANGES.md [hunk 12]
```diff
@@ -5,10 +5,18 @@
...
 
 ##### Additions :tada:
 
 - Added a constructor overload for `Cesium3DTilesSelection::ITwinRealityDataContentLoaderFactory` to override the iTwin Cesium Reality Data base URL. This makes it possible to connect to alternate servers (e.g., staging, QA, mock servers).
+- Added `PositionFromAccessor` visitor for reading positions from a `PositionAccessorType`.
+- Added `NormalFromAccessor` visitor for reading normals from a `NormalAccessorType`.
+- Added `QuaternionFromAccessor` visitor for reading quaternions from a `QuaternionAccessorType`.
+- Updated `TexCoordAccessorType` to support additional quantized variants enabled by `KHR_mesh_quantization`.
+- Added `AccessorView::normalized()` and `AccessorWRiter::normalized()` getters.
```
Comment (j9liu) [inline CHANGES.md]:
```suggestion
- Added `AccessorView::normalized()` and `AccessorWriter::normalized()` getters.
```

File: CesiumGltf/test/TestAccessorUtility.cpp [hunk 13]
```diff
@@ -1074,9 +1370,273 @@ TEST_CASE("Test TexCoordFromAccessor") {
...
+    REQUIRE(
+        std::visit(StatusFromAccessor{}, quaternionAccessor) ==
+        AccessorViewStatus::Valid);
+    REQUIRE(
+        std::visit(CountFromAccessor{}, quaternionAccessor) ==
+        quaternions.size());
+  }
+}
+
+TEST_CASE("Test PositionFromAccessor") {
```
Comment (j9liu) [inline CesiumGltf/test/TestAccessorUtility.cpp]:
```suggestion
TEST_CASE("Test QuaternionFromAccessor") {
```

File: CesiumGltf/test/TestAccessorUtility.cpp [hunk 14]
```diff
@@ -941,19 +1222,6 @@ TEST_CASE("Test getTexCoordAccessorView") {
     model.accessors[0].componentType = Accessor::ComponentType::FLOAT;
   }
 
-  SUBCASE("Handles invalid un-normalized texcoord") {
```
Comment (j9liu) [inline CesiumGltf/test/TestAccessorUtility.cpp:944]: What's the reason for removing this? Don't the accessors have to report that they're normalized in order to be valid?

**Review by j9liu — COMMENTED** [review]
Thanks @lilleyse ! Most of my comments are doc-related, there aren't any major changes I requested.

I did want to verify that things still work in Unreal/Unity after this fix, will do that on my end before merge.

File: CesiumGltf/src/AccessorUtility.cpp (hunk 11 above)
Comment (lilleyse) [reply CesiumGltf/src/AccessorUtility.cpp:36]: It depends on the attribute. `NORMAL` attributes must be normalized whereas `POSITION` attributes may be normalized or unnormalized. The full rules are here:

glTF mesh|glTF animation samplers|`KHR_mesh_quantization`|`EXT_mesh_gpu_instancing`
--|--|--|--
image|image|image|image

File: CesiumGltf/test/TestAccessorUtility.cpp (hunk 14 above)
Comment (lilleyse) [reply CesiumGltf/test/TestAccessorUtility.cpp:944]: > Don't the accessors have to report that they're normalized in order to be valid?

That's true for the core glTF spec, but `KHR_mesh_quantization` allows them to be unnormalized.

**Comment by lilleyse** [comment]
Thanks for the suggestions @j9liu, should be addressed now. Ready for another review.

**Review by j9liu — APPROVED** [review]
Thanks @lilleyse !
